Convert openshift-sdn to a CNI plugin

dcbw opened the pull request:
@danwinship see how you like this one better; the CNI plugin is now essentially a small HTTP client that sends the CNI config and environment over a root-owned unix domain socket to the openshift-node process which serializes all the pod requests and does the actual network setup work.  Hostports not implemented yet, but they'll be less icky than the other branch since we can do it all in openshift-node rather than having to work around that race.

dcbw commented:
@danwinship it's based off your kill-registry branch, obviously... ignore those commits.

danwinship commented:
I assume the IPAM commit is the same as before?

danwinship commented on pkg/sdn/plugin/node.go:
```diff
@@ -120,18 +128,74 @@ func (node *OsdnNode) Start() error {
-			err = node.UpdatePod(p.Namespace, p.Name, kubeletTypes.ContainerID{ID: containerID})
 			if err != nil {
+				err = node.UpdatePod(p.Namespace, p.Name, kubeletTypes.ContainerID{ID: containerID})
 				log.Warningf("Could not update pod %q (%s): %s", p.Name, containerID, err)
```
accidental change?

dcbw replied:
Yep, fixed.

danwinship commented on pkg/sdn/plugin/pod.go:
```diff
@@ -0,0 +1,410 @@
+
+	AssignMacVlanAnnotation string = "pod.network.openshift.io/assign-macvlan"
+
+	interfaceName = knetwork.DefaultInterfaceName
```
"podInterfaceName"?

dcbw replied:
Done.

danwinship commented on pkg/sdn/plugin/pod.go:
```diff
@@ -0,0 +1,410 @@
+	"github.com/containernetworking/cni/pkg/ip"
+	"github.com/containernetworking/cni/pkg/ipam"
+	"github.com/containernetworking/cni/pkg/ns"
+	"github.com/containernetworking/cni/pkg/types"
```
maybe import as "cnitypes" for clarity

dcbw replied:
done

danwinship commented on pkg/sdn/plugin/pod.go:
```diff
@@ -0,0 +1,410 @@
+					IP:   net.IPv4zero,
+					Mask: net.IPMask(net.IPv4zero),
+				},
+					GW: netutils.GenerateDefaultGateway(nodeNet)},
```
This indentation is terrible... gofmt accepts it, but... ugh. There must be some better way.

gofmt also accepts it if you put a newline between "{" and "Dst" and then indents Dst and GW at the same level

dcbw replied:
Yeah ugly, it was gofmt -s -w's suggestion but I've taken your suggestion about newlines.

stevekuznetsov commented on hack/common.sh:
```diff
@@ -294,6 +290,18 @@ os::build::internal::build_binaries() {
+      local -a tests=()
+      for binary in "${binaries[@]}"; do
+        if [[ "${binary}" =~ ".test"$ ]]; then
+          tests+=($binary)
```
```sh
tests+=("${binary}")
        ^ ^      ^^
```

dcbw replied:
Fixed.

stevekuznetsov commented on hack/common.sh:
```diff
@@ -336,10 +344,13 @@ os::build::internal::build_binaries() {
+# a for given platform.  First argument is platform, remaining arguments are
+# targets.  Targets can be given as full Go package path or as basenames.
+function os::build::export_targets_and_binaries() {
+  local platform=${1}
```
We prefer not to brace positional args smaller than `10`. We do prefer to quote, though.

dcbw replied:
Fixed.

stevekuznetsov commented on hack/common.sh:
```diff
@@ -349,17 +360,26 @@ function os::build::export_targets() {
+# Generates the set of target platforms to build for. Accepts platforms via
+# OS_BUILD_PLATFORMS, or defaults to the current platform.
+function os::build::export_platforms() {
   platforms=("${OS_BUILD_PLATFORMS[@]:+${OS_BUILD_PLATFORMS[@]}}")
```
```sh
platforms=("${OS_BUILD_PLATFORMS[@]:+"${OS_BUILD_PLATFORMS[@]}"}")
                                     ^                        ^
```

dcbw replied:
Fixed.

stevekuznetsov commented on origin.spec:
```diff
@@ -267,12 +267,16 @@ mkdir -p %{buildroot}%{_sharedstatedir}/origin
+install -d -m 0755 %{buildroot}/opt/cni/bin
+install -p -m 0755 _build/bin/sdn-cni-plugin %{buildroot}/opt/cni/bin/openshift-sdn
+install -p -m 0755 _build/bin/host-local %{buildroot}/opt/cni/bin
+install -p -m 0755 _build/bin/loopback %{buildroot}/opt/cni/bin
```
My personal preference is always for the verbose mode specifiers since you can reason about them without memorizing arcane sets of bits.

dcbw replied:
Yeah, though I opted to use the existing style.  Is that OK?

stevekuznetsov replied:
I'll deal with it :)

eparis replied:
/me prefers the numbers, so much easier to read   :)

stevekuznetsov replied:
I'll blame it on my age

smarterclayton commented on pkg/sdn/plugin/node.go:
```diff
@@ -120,18 +128,74 @@ func (node *OsdnNode) Start() error {
+func (node *OsdnNode) watchCni() {
+	for {
+		select {
+		case request := <-node.podRequestChan:
```
It looks like this loop can block if any CNI container fails.  You'd need to have independent pod queues and handle them separately, and deal with CNI clients that crash and can't receive their updates.  It would be better to preserve all of the updates for each pod in a local cache and identify which to serve to the client (the latest).

dcbw replied:
@smarterclayton did that in latest pushes.

dcbw commented on pkg/sdn/plugin/cniserver.go:
```diff
@@ -0,0 +1,213 @@
+type CniCommand string
+
+const CNI_ADD CniCommand = "ADD"
+const CNI_UPDATE CniCommand = "UPDATE"
```
It's mostly to ensure requests are serialized; eg that an UPDATE cannot happen concurrently with an ADD.  Does that make more sense?

danwinship replied:
Hm... actually I think I misunderstood the code before; updates are not actually passing through the openshift-sdn CNI binary, right? It goes watchNetNamespaces() -> node.go:UpdatePod() -> podRequestChan -> watchCni() -> pod.go:updatePod(). OK.

dcbw replied:
@danwinship correct; UPDATE actions are not passing through the CNI plugin, they are just directly funneled into the CNI event queue on the node process.

dcbw commented on pkg/sdn/plugin/node.go:
```diff
@@ -70,10 +75,13 @@ func NewNodePlugin(pluginName string, osClient *osclient.Client, kClient *kclien
 		kClient:            kClient,
 		osClient:           osClient,
+		cniServer:          NewCniServer(podChan),
+		podRequestChan:     podChan,
```
Then we couldn't pass that channel to NewCniServer() though.  It needs to be the same channel, hence why it got broken out.

dcbw commented:
> I assume the IPAM commit is the same as before?

Yes, just rebased.

dcbw commented on pkg/sdn/plugin/cniserver.go:
```diff
@@ -0,0 +1,213 @@
+	namespace string
+	name      string
+	id        string
+	netns     string
```
I think all usage of 'namespace' in the IPAM commit is actually the namespace name.  The stuff that's kernel-netns related is called "netns".

dcbw replied:
Changed podRequest struct fields to be clearer.

dcbw commented on pkg/sdn/plugin/cniserver.go:
```diff
@@ -0,0 +1,213 @@
+	}
+}
+
+func (s *CniServer) cniDelete(w http.ResponseWriter, r *http.Request) {
```
Fixed.

dcbw commented on pkg/sdn/plugin/node.go:
```diff
@@ -1,47 +1,52 @@
 	localIP            string
 	localSubnet        *osapi.HostSubnet
 	hostName           string
+	hostSubnet         *osapi.HostSubnet
```
removed.

dcbw commented:
@danwinship @openshift/networking @smarterclayton @stevekuznetsov rebased and fixed up, PTAL thanks!

danwinship commented on origin.spec:
```diff
@@ -267,9 +267,16 @@ mkdir -p %{buildroot}%{_sharedstatedir}/origin
+install -d -m 0755 %{buildroot}/opt/cni/bin
+install -p -m 0755 _build/bin/sdn-cni-plugin %{buildroot}/opt/cni/bin/openshift-sdn
+install -p -m 0755 _build/bin/host-local %{buildroot}/opt/cni/bin
+install -p -m 0755 _build/bin/loopback %{buildroot}/opt/cni/bin
```
What is /opt/cni/bin/loopback?

dcbw replied:
@danwinship it's used by the kubelet CNI driver to configure the loopback interface inside the container.  CNI plugins don't have to do that themselves.

danwinship commented on pkg/cmd/server/kubernetes/node_config.go:
```diff
@@ -190,6 +185,23 @@ func BuildKubernetesNodeConfig(options configapi.NodeConfig, enableProxy, enable
 		return nil, err
 	}
 
+	// Initialize SDN before building kubelet config so it can modify options
```
"so it can modify options"? how so?

danwinship replied:
ok, so the rearrangement from how it was before is not actually needed, it's just aesthetic? I guess the old way keeps all of the `server` initialization code together, while the new way keeps all the SDN code together...

smarterclayton replied:
We want all of the server initialization code together - that's a commonly reviewed bit of code, and if someone comes in here now they will miss it.

smarterclayton replied:
Does NewNodePlugin start goroutines?  If it does, it cannot be in this method, or you need to split the starting goroutines part into a `Start()/Run()` method that is called later.

danwinship commented on pkg/sdn/plugin/node.go:
```diff
@@ -127,11 +139,138 @@ func (node *OsdnNode) Start() error {
+		}
+		request := w.requests[0]
+		w.requests = w.requests[1:]
+		w.lock.Unlock()
```
ugh. no! you're just reinventing channels

If you want to avoid possibly blocking on channel send, you can just do the send from a goroutine.

Also, what is all of this complexity (relative to your original version) protecting against? setupPod/teardownPod/updatePod hanging forever? Would the old code have dealt with that? Can't we just use some sort of timeout+error instead of having multiple workers?

dcbw replied:
@smarterclayton requested this in https://github.com/openshift/origin/pull/11082#discussion_r80743524 if I understand his request correctly.  And he's at least right from the perspective that if pod setup goes into the weeds for a single pod, that would (in the original push) block any subsequent pods from being created.

dcbw replied:
Yeah, it kinda reinvents channels, but with two important differences: (1) it's non-blocking and we don't have to specify a channel capacity (after which things would block) and (2) we can cleanly terminate the pod worker when it no longer has work to do which turns out to be pretty racy with a channel.  (1) is not much of a problem (though annoying) and (2) is harder to solve, but I'll see if I can come up with something not-ugly.

dcbw replied:
I fixed this up and it's not as bad as I thought.  PTAL?

stevekuznetsov replied:
>  it's non-blocking 

If you don't care that someone is there to get the message (otherwise you'd block, right?) you could always:
```go
select {
case myChan <- myVal:
case <- time.After(5 * time.Second)
}
```

> which turns out to be pretty racy with a channel

On the surface this seems like a misuse of concurrency primitives.

danwinship commented on pkg/sdn/plugin/node.go:
```diff
@@ -127,11 +139,138 @@ func (node *OsdnNode) Start() error {
+		podNamespace: namespace,
+		podName:      name,
+		containerId:  id.String(),
+		netns:        "/blah/something", // plugin doesn't care about namespace
```
Comment doesn't appear to be true; pod.go:updatePod() calls `getVethInfo(req.netns, "eth0")` which looks like it will return an error if netns is bogus.

dcbw replied:
Fixed.

danwinship commented on pkg/sdn/plugin/pod.go:
```diff
@@ -0,0 +1,412 @@
+}
+
+// Returns host veth, container veth MAC, and pod IP
+func getVethInfo(netns, containerIfname string) (netlink.Link, string, string, error) {
```
I'd commented before about how the comment claimed it returned the veth name but it actually returned the whole netlink.Link, and you changed the comment, but I'd meant to change the return value, because all of its callers only use the name, not anything else

dcbw replied:
Fixed.

dcbw commented on pkg/sdn/plugin/controller.go:
```diff
@@ -320,17 +254,6 @@ func (plugin *OsdnNode) SetupSDN(localSubnetCIDR, clusterNetworkCIDR, servicesNe
-	// (This has to have been performed in advance for docker-in-docker deployments,
-	// since this will fail there).
-	_, _ = exec.Command("modprobe", "br_netfilter").CombinedOutput()
-	err = sysctl.SetSysctl("net/bridge/bridge-nf-call-iptables", 0)
```
Fixed.

dcbw commented:
@danwinship @smarterclayton PTAL, thanks!

danwinship commented:
still need to figure out if /run/openshift-sdn/config.env is needed

dcbw commented:
Rebased and adjusted for dind changes.  @rajatchopra do you still need /run/openshift-sdn/config.env for anything in the router?

rajatchopra commented:
@dcbw yes we still need the config.env for F5 automation scripts for F5 versions less than 12.1

smarterclayton commented on pkg/sdn/plugin/node.go:
```diff
@@ -125,18 +149,140 @@ func (node *OsdnNode) Start() error {
+	return &podWorker{
+		node: node,
+		// Allow queue depth of 10 before blocking additions, if there
+		// are more than 10 outstanding requests something is wrong
```
How are you going to detect something is wrong?

dcbw replied:
Completely reworked this code.

smarterclayton commented on pkg/sdn/plugin/node.go:
```diff
@@ -125,18 +149,140 @@ func (node *OsdnNode) Start() error {
+
+func (w *podWorker) addRequest(request *podRequest) {
+	w.requests <- request
+	if len(w.requests) == 1 {
```
What are you trying to do here?  If you're trying to be clever and only run a goroutine while requests are pending that's not really going to make this more efficient.  Also, is this called within a lock / single threaded context?  If not this code is racy.

smarterclayton commented on pkg/sdn/plugin/node.go:
```diff
@@ -125,18 +149,140 @@ func (node *OsdnNode) Start() error {
+	return fmt.Sprintf("%s/%s", request.podNamespace, request.podName)
+}
+
+func (node *OsdnNode) watchCni() {
```
methods like this are almost always called "run" or "handle" some variant.  I would suggest you call this `handleCNIRequests` and you should also provide thorough Godoc describing the loop you are adding here.

dcbw replied:
Done.

smarterclayton commented on pkg/sdn/plugin/node.go:
```diff
@@ -125,18 +149,140 @@ func (node *OsdnNode) Start() error {
+
+	// Grab the netns from our cache since the Update request doesn't come from CNI
+	node.runningPodsLock.Lock()
+	runningPod, ok := node.runningPods[getPodKey(req)]
```
Make runningPods its own type with safe accessors.  Multiple locks per type complicates reasoning about the effects of the lock and is generally better represented by encapsulating the locks in function calls. 

In general, if you have `.Unlock()` and it's not in a defer, you should rearrange your code until the lock and object under the lock until they are inside a function.

dcbw replied:
runningPods is now only accessed from a single goroutine and no longer needs a lock.

smarterclayton commented on pkg/sdn/plugin/node.go:
```diff
@@ -125,18 +149,140 @@ func (node *OsdnNode) Start() error {
+
+// FIXME: this should eventually go into kubelet via a CNI UPDATE/CHANGE action
+// See https://github.com/containernetworking/cni/issues/89
+func (node *OsdnNode) UpdatePod(namespace string, name string, id kubeletTypes.ContainerID) error {
```
All of the function here needs a good unit test that validates this fairly complicated new logic works, especially in the presence of multiple pods and goroutines adding and acting on the objects at once.

I'd recommend a few basic unit tests validating the core logic, then a more sophisticated test that adds a set of mutations for different objects in multiple goroutines and validates everything comes out the other end.

dcbw replied:
Added unit tests for the cniserver, the actual CNI plugin that kubelet calls, and the internal pod queue/worker code.

marun commented:
dind changes lgtm

dcbw commented on pkg/sdn/plugin/node.go:
```diff
@@ -125,18 +149,140 @@ func (node *OsdnNode) Start() error {
+	req.netns = runningPod.netns
+
+	// Send request and wait for the result
+	node.podRequestChan <- req
```
Done, it's all now in the podWorker type, which helps facilitate testing.

dcbw commented on pkg/sdn/plugin/pod.go:
```diff
@@ -0,0 +1,453 @@
+package plugin
```
Lots of godoc added.

smarterclayton replied:
Please follow the godoc conventions - always start with the method name, complete sentences.  General style applies to all our code.  Interfaces should be documented as well.

Field godoc should be `// fieldName does blah to blarg`

dcbw commented:
@danwinship @smarterclayton @openshift/networking mega-update.

This is much simplified from the previous push in that it serializes all pod operations (like we currently do, because of the locking in pkg/sdn/plugin/bin/openshift-sdn-ovs) instead of the previous push that tried to serial all operations *per pod*.  That was problematic (and the reason this re-push took so long since I tried to solve these before giving up) because it seems to be really hard to GC the pod workers (without adding another global pod watch) in a race-free manner when using go channels since incoming requests from the HTTP server happen on a goroutine, and we need to wait for the pod worker to be done with the current operation before killing it.

Anyway, new push should be simpler, we already serialized pod operations before, and there are testcases for the new code and the new CNI plugin.

dcbw commented:
@danwinship can you double-check the iptables thing in the last commit?  I tried to test it out last week, but could reproduce the failure we were talking about with the original code either.

dcbw commented:
@smarterclayton @openshift/networking important to note that docker will no longer show the pod's IP address in 'docker inspect' since we are no longer using docker for IPAM.  This clearly needs a release note somewhere, and it's been discussed before via a Tech Talk, internal AOS team communications, and with PM.  But not too much externally.

dcbw commented:
[testextended][extended:networking]

dcbw commented:
Sigh; I need to fix "vendor/github.com/vishvananda/netlink/netlink.go:14:2: no buildable Go source files in /home/dcbw/Development/containers/origin/_output/local/go/src/github.com/openshift/origin/vendor/github.com/vishvananda/netlink/nl" for non-Linux builds, since openshift-sdn now uses netlink and since it's part of the openshift binary it gets built on Darwin/Windows too...

dcbw commented:
[testextended][extended:networking]

dcbw commented:
Travis errors are govet variable shadowing, flake https://github.com/openshift/origin/issues/6852

[test]

dcbw commented:
[testextended][extended:networking]

danwinship commented:
> This clearly needs a release note somewhere

openshift-docs #2933

danwinship commented:
Hm... In the 1.3 -> 1.4 upgrade process, we need to make sure somehow that docker gets restarted (after we have removed docker-sdn-ovs.conf); that will both ensure that docker is back to using docker0/172.17.0.0 rather than the SDN subnet, and that any running pods get killed and restarted and re-IPAM'ed so that you don't have pods using IPs that CNI IPAM doesn't know about. And ideally, we would clean up lbr0/vovsbr/vlinuxbr too. Oh, and we need to delete any old iptables rules referring to the SDN subnet that docker leaves behind after restarting.

Anything else?

danwinship commented:
> @danwinship can you double-check the iptables thing in the last commit? I tried to test it out last week, but could reproduce the failure we were talking about with the original code either.

Test it by having a pod try to connect to a service that is not backed by the same pod. (Kubernetes installs its own masquerade rule to deal with the pod-to-self case.) You'll have a pod IP connecting to a service IP, which will get routed to iptables by OVS, and iptables will rewrite the service IP to the server pod IP. With the original rule:

    -		{"nat", "POSTROUTING", []string{"-s", n.clusterNetworkCIDR, "!", "-d", n.clusterNetworkCIDR, "-j", "MASQUERADE"}},

this packet doesn't match, so it doesn't get masqueraded, so the packet will get delivered to the server pod with its original source IP, so the server will attempt to respond directly to the client pod, and the client pod will reject the response because it was expecting a response from the service IP, not the server pod IP.

With the new rule:

    +		{"nat", "POSTROUTING", []string{"-s", n.clusterNetworkCIDR, "-j", "MASQUERADE"}},

the packet does get masqueraded, so the server pod response will be delivered back to the node, which can then unmasquerade it and forward it to the client pod with the expected IP.

(Tested. It works for me with the new rule, but not with the old rule.)

knobunc commented:
@smarterclayton @stevekuznetsov Please re-review.

knobunc commented on hack/common.sh:
```diff
@@ -301,6 +297,18 @@ os::build::internal::build_binaries() {
+      local -a nonstatics=()
+      local -a tests=()
+      for binary in "${binaries[@]}"; do
+        if [[ "${binary}" =~ ".test"$ ]]; then
```
A bare $ always scares me with bash.  Should we just say:
 =~ '.test$'

stevekuznetsov replied:
That's not equivalent -- `'.test'$` will test that `'.test'` is followed by a line end anchor, `'.test$'` will test that there is a literal `'$'` after `'.test'`.

This is pre-existing code so I don't think we ought to change it here, but seeing stuff like this makes me feel like there is something wrong with the higher level logic, that `$binaries` list is clearly two lists in one.

knobunc replied:
Wow... right you are!  Tricky bash...

I stopped reading at "When it is used, the string to the right of the operator is considered an extended regular expression and matched accordingly (as in regex(3))." and didn't see the bit about quoting special characters.

stevekuznetsov replied:
Fun stuff, huh? :)

knobunc commented:
@liggitt would you have time to cast an eye over these changes please?

smarterclayton commented on hack/common.sh:
```diff
@@ -356,17 +367,26 @@ function os::build::export_targets() {
   if [[ ${#targets[@]} -eq 0 ]]; then
-    targets=("${OS_ALL_TARGETS[@]}")
+    if [[ "${platform}" == linux/amd64 ]]; then
+      targets=("${OS_ALL_TARGETS[@]}" "${OS_SDN_COMPILE_TARGETS_LINUX_AMD64[@]}")
```
This is wrong - the abstraction before was that you didn't have to know this info in here.  If you want to build OS_SDN_COMPILE_TARGETS, pass them in to this method.

smarterclayton replied:
Or explain more clearly why you wanted this, because I don't see how this makes the method more correct.

dcbw replied:
@smarterclayton the logic is now in hack/build-go.sh, which is the only place that passes an empty targets list to os::build::build_binaries().  All other users correctly pass either OS_CROSS_COMPILE_TARGETS or OS_IMAGE_COMPILE_TARGETS.

build-go.sh gets used from a ton of places that also need the SDN components built, so it now gets the logic.

smarterclayton commented on pkg/sdn/plugin/cniserver/cniserver.go:
```diff
@@ -0,0 +1,212 @@
+	CNI_DEL:    "/cni/del",
+}
+
+type CniRequest struct {
```
All of these should be `CNIRequest`.

smarterclayton replied:
Since it's an acronym

dcbw replied:
Done, and all the other CNI references.

smarterclayton commented on pkg/sdn/plugin/pod.go:
```diff
@@ -0,0 +1,190 @@
+}
+
+// Creates a new basic podManager; used by testcases
+func newPodManagerCommon() *podManager {
```
`newDefaultPodManager()`

dcbw replied:
Done

smarterclayton commented on pkg/sdn/plugin/pod_linux.go:
```diff
@@ -0,0 +1,401 @@
+	statusCmd   = "status"
+	updateCmd   = "update"
+
+	AssignMacVlanAnnotation string = "pod.network.openshift.io/assign-macvlan"
```
`MACVLAN`

dcbw replied:
Sure, can change.  This was a C&P from other code though.

smarterclayton commented on pkg/sdn/plugin/poddb.go:
```diff
@@ -0,0 +1,46 @@
+package plugin
```
Who cleans this up on restart of a node?

danwinship replied:
Yeah... I don't really like this. We should just pass the containerRuntime to the node so it can look up namespaces when it needs to. Yes, we'll need to do something else when we split the SDN out of the node process, but there are lots of other things we'll need to do something about as well, and this is no worse.

(The sync-all-pod-network-states-on-startup loop probably belongs in upstream kubernetes anyway; if that happened then we'd get an UPDATE-with-netns-filled-in for every pod at startup and then we'd be guaranteed to have a netns for every pod after that.)

dcbw replied:
@danwinship there's no good way to pull the runtime out of kubelet since openshift doens't even get the kubelet object. The best we can do is continue to provide our own kubelet network plugin, but as a proxy for the actual kubelet CNI driver, and then save the Host in our Init() call before passing the init to CNI.

So we take a step backwards and can no longer use the normal kubelet CNI plugin, but I guess it works for now.  See how you like the changes I pushed, especially pkg/sdn/plugin/plugin.go.

dcbw commented on pkg/sdn/plugin/sdn-cni-plugin/openshift-sdn.go:
```diff
@@ -0,0 +1,115 @@
+
+// Send the DEL command environment and config to the CNI server
+func (p *cniPlugin) CmdDel(args *skel.CmdArgs) error {
+	_, err := p.doCni(fmt.Sprintf("http://d%s", cniserver.CNI_URLS[cniserver.CNI_DEL]), newCniRequest(args))
```
@smarterclayton following @danwinship suggestion, this is now a static string and no longer Sprintf-ed.

dcbw commented on pkg/sdn/plugin/cniserver/cniserver.go:
```diff
@@ -0,0 +1,212 @@
+package cniserver
```
Done.

dcbw commented on pkg/sdn/plugin/cniserver/cniserver.go:
```diff
@@ -0,0 +1,212 @@
+
+type cniRequestFunc func(request *PodRequest) ([]byte, error)
+
+type CniServer struct {
```
@smarterclayton done.

dcbw commented:
travis flake is https://github.com/openshift/origin/issues/9457 [test]

dcbw commented:
> Hm... In the 1.3 -> 1.4 upgrade process, we need to make sure somehow that docker gets restarted (after we have removed docker-sdn-ovs.conf);

@danwinship updgrade logic is now in controller.go, dockerCleanup().  Basically moving the systemd restart logic into controller.go from openshift-sdn-docker-setup.sh.

dcbw commented:
@danwinship note that with the new cleanup logic, there are still some rules that reference lbr0:
-A FORWARD -o lbr0 -j DOCKER
-A FORWARD -o lbr0 -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT
-A FORWARD -i lbr0 ! -o lbr0 -j ACCEPT
-A FORWARD -i lbr0 -o lbr0 -j ACCEPT

but these should be harmless as they require either the in/out interface to be lbr0, which no longer exists, thus they should never be hit.  I opted to just leave them.

dcbw commented:
More shadowing errors, flake https://github.com/openshift/origin/issues/6852 re-[test]

knobunc commented:
[test]

knobunc commented:
@smarterclayton @stevekuznetsov do the changes address your concerns?

danwinship commented on pkg/sdn/plugin/controller.go:
```diff
@@ -176,85 +178,84 @@ func deleteLocalSubnetRoute(device, localSubnetCIDR string) {
+	interfaces := []string{"lbr0", "vovsbr", "vlinuxbr"}
+
+	exec := kexec.New()
+	for _, intf := range interfaces {
```
You don't need to test all three; we only need to handle transitions from working configurations.

danwinship commented on pkg/sdn/plugin/controller.go:
```diff
@@ -176,85 +178,84 @@ func deleteLocalSubnetRoute(device, localSubnetCIDR string) {
+	if err := osexec.Command("dbus-send", "--system", "--print-reply", "--reply-timeout=2000", "--type=method_call", "--dest=org.freedesktop.systemd1", "/org/freedesktop/systemd1", "org.freedesktop.systemd1.Manager.Reload"); err != nil {
+		glog.Error(err)
 	}
+	if err := osexec.Command("dbus-send", "--system", "--print-reply", "--reply-timeout=2000", "--type=method_call", "--dest=org.freedesktop.systemd1", "/org/freedesktop/systemd1", "org.freedesktop.systemd1.Manager.RestartUnit", "string:'docker.service' string:'replace'"); err != nil {
```
So this brings back the "we restart docker but then don't wait for it to finish restarting" bug that was otherwise going to go away, at least in the upgrade case. Consider an ExponentialBackoff checking "docker ps" or something to detect when it's running again?

knobunc commented:
[test]

stevekuznetsov commented on contrib/node/install-sdn.sh:
```diff
@@ -1,36 +1,31 @@
 #!/bin/bash
+set -x
```
Is this supposed to be here?

dcbw replied:
It sure isn't.

stevekuznetsov commented on contrib/node/install-sdn.sh:
```diff
@@ -1,36 +1,31 @@
-  if [ ! -d ${target} ]; then
-    mkdir -p ${target}
-  fi
+  mkdir -p -m 0755 "${target_usrdir}"
```
I think I commented on these before but I still think `o+rwx,g+rx,o+rx` is infinitely more readable (also, why do we need to be this explicit? We don't want `g+w`?).

dcbw replied:
@stevekuznetsov I'm OK changing this, but this format is used nowhere else in OpenShift scripts that I can see, nor in origin.spec, nor in Golang calls like os.Mkdir() and os.Chmod().  We can certainly blaze the trail here, at the cost of inconsistency.

stevekuznetsov replied:
Yeah, that is what I was afraid of. That level of inconsistency does not upset me much, but let's put this off for now. I just don't like asking everyone that ever reads this script to do the mental mapping from `0755` to binary and then to the permissions when we could do that mapping once, encode it in the script in the manner in which people actually want to read it, and call it a day.

knobunc replied:
Personally I find the the numbers to be far clearer...

stevekuznetsov replied:
As above I'll blame it on my age. `o+rx` == others can read and execute. `5` == ???? :)

stevekuznetsov commented on contrib/vagrant/provision-util.sh:
```diff
@@ -49,7 +49,7 @@ os::provision::base-install() {
   echo "Installing openshift"
   os::provision::install-cmds "${origin_root}"
-  os::provision::install-sdn "${origin_root}"
+  os::provision::install-sdn "${origin_root}" $(os::build::get-bin-output-path "${OS_ROOT}")
```
```sh
os::provision::install-sdn "${origin_root}" "$(os::build::get-bin-output-path "${OS_ROOT}")"
                                            ^                                              ^
```

stevekuznetsov commented on hack/build-go.sh:
```diff
@@ -4,11 +4,13 @@
 source "$(dirname "${BASH_SOURCE}")/lib/init.sh"
 
+build_targets="$@"
+platform=$(os::build::host_platform)
```
```sh
platform="$(os::build::host_platform)"
         ^                           ^
```

smarterclayton commented on pkg/sdn/plugin/pod.go:
```diff
@@ -0,0 +1,193 @@
+		}
+		glog.V(5).Infof("Processed pod network request %v, result %s err %v", request, string(result.Response), result.Err)
+		request.Result <- result
+	}
```
Unless you're adding a Close method, you should panic if you get here because that means this process is no longer managing pod updates (which is bad)

knobunc commented:
re-[test] flaked on https://github.com/openshift/origin/issues/11366

dcbw commented on pkg/sdn/plugin/pod.go:
```diff
@@ -0,0 +1,194 @@
+			result.Err = m.podHandler.update(request)
+		case cniserver.CNI_DEL:
+			delete(m.runningPods, pk)
+			result.Err = m.podHandler.teardown(request)
```
Done and testcase added.

eparis commented:
[merge] get in the queue!

dcbw commented:
@smarterclayton to address your points (instead of irc):

1. what are the remaining risks for CNI?
User surprise at pod IPs no longer showing up in 'docker inspect' which I've addressed through a docs change, and of course the new codepaths that the CNI code takes that haven't been used before in openshift (but have in upstream kubernetes)

2. if we merge this for 1.4, what are our fallback positions if we hit something unexpected that causes hangs, blocks, etc?
We're committed to fixing this, and if it's not fixable within the time required, we'll either (a) back it out, or (b) disable the CNI plugin/server component and effectively revert to the old flow without reverting the entire PR.

3. have you done enough stress testing to feel confident this can't hang?
Yes, I have done pod create/delete stress testing with 50+ pods and the only issues I've seen are caused by docker timeouts and its inability to handle more than 20+ pods on the same node without 'docker ps' hanging itself from a tree.  But the network setup/teardown completed successfully in these cases.  It's also completed the extended network tests successfully.

4. have you reviewed the logic several times for concurrency bugs?  yes, I've reviewed and reworked the multiple times in the last weeks to simplify things greatly based largely on your quite useful feedback.  @danwinship has also reviewed for concurrency issues.  We believe it is robust and test coverage is greatly increased for both old and new code.

dcbw commented:
If anything, this might be *more* robust than docker, allowing us to avoid stuff like:

Oct 21 20:08:04 openshift-node-2 docker[9512]: sync.(*Mutex).Lock(0xc8205e88b4)
Oct 21 20:08:04 openshift-node-2 docker[9512]:         /usr/lib/golang/src/sync/mutex.go:83 +0x1c4 fp=0xc8216546e8 sp=0xc821654698
Oct 21 20:08:04 openshift-node-2 docker[9512]: github.com/docker/libnetwork/datastore.(*datastore).List(0xc8205e8880, 0xc8220d6200, 0x1c, 0x7fe23b4131d0, 0xc823306b40, 0x0, 0x0, 0x0, 0x0, 0x0)
Oct 21 20:08:04 openshift-node-2 docker[9512]:         /builddir/build/BUILD/docker-e03ddb80cc2e3244b70052ec38d29cf526444c59/vendor/src/github.com/docker/libnetwork/datastore/datastore.go:468 +0x6f fp=0xc8216548
Oct 21 20:08:04 openshift-node-2 docker[9512]: github.com/docker/libnetwork.(*controller).getNetworksFromStore(0xc8201666e0, 0x0, 0x0, 0x0, 0x0, 0x0)
Oct 21 20:08:04 openshift-node-2 docker[9512]:         /builddir/build/BUILD/docker-e03ddb80cc2e3244b70052ec38d29cf526444c59/vendor/src/github.com/docker/libnetwork/store.go:124 +0x25a fp=0xc821654a70 sp=0xc8216

dcbw commented:
More CI flake: can't load package: "package tools/genman: cannot find package "tools/genman" in any of:"

dcbw commented:
Repush to fix "test/integration/sdn_test.go:14: imported and not used: "github.com/openshift/origin/pkg/sdn/plugin" as sdnplugin"

dcbw commented:
Repushed to fix a hack/build-go.sh issue with integration tests

dcbw commented:
[test] [testextended][extended:networking]

dcbw commented:
origin_integration test failed with flake (probably DDOS related?)

```
Running TestRegistryClientAPIv1...
failed  TestRegistryClientAPIv1
I1021 18:54:54.172192   26825 client.go:409] Getting repository openshift/origin-not-found from {https  <nil> registry.hub.docker.com    }
I1021 18:54:54.280093   26825 client.go:409] Getting repository openshift/schema-v1-test-repo from {https  <nil> registry.hub.docker.com    }
--- FAIL: TestRegistryClientAPIv1 (0.17s)
	dockerregistryclient_test.go:246: openshift/schema-v1-test-repo was pushed by Docker 1.8.2
	dockerregistryclient_test.go:207: error retrieving repository: server returned 500
```

dcbw commented:
conformance test flake:
```
• Failure in Spec Teardown (AfterEach) [300.193 seconds]
[k8s.io] LimitRange
/data/src/github.com/openshift/origin/vendor/k8s.io/kubernetes/test/e2e/framework/framework.go:793
  should create a LimitRange with defaults and ensure pod has those defaults applied. [AfterEach]
  /data/src/github.com/openshift/origin/vendor/k8s.io/kubernetes/test/e2e/limit_range.go:102

  Oct 21 18:50:24.610: Couldn't delete ns: "e2e-tests-limitrange-3h2at": namespace e2e-tests-limitrange-3h2at was not deleted with limit: timed out waiting for the condition, namespace is empty but is not yet removed (&errors.errorString{s:"namespace e2e-tests-limitrange-3h2at was not deleted with limit: timed out waiting for the condition, namespace is empty but is not yet removed"})

  /data/src/github.com/openshift/origin/vendor/k8s.io/kubernetes/test/e2e/framework/framework.go:338
------------------------------

```

danwinship commented:
flake #11094 (plus also apparently an infrastructure-related flake involving the docker hub). [test]

eparis commented:
[merge]

dcbw commented:
[test]

eparis commented:
might need another [merge] I never remember

dcbw commented:
travis flake:
```
INFO] Updating /home/travis/gopath/src/github.com/openshift/origin/_output/verify-generated-swagger-spec/api/swagger-spec/oapi-v1.json from https://127.0.0.1:38443/swaggerapi/oapi/v1...

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current

                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0curl: (35) error:1407742E:SSL routines:SSL23_GET_SERVER_HELLO:tlsv1 alert protocol version

[ERROR] PID 23055: /home/travis/gopath/src/github.com/openshift/origin/hack/update-generated-swagger-spec.sh:50: `curl -w "\n" "${SWAGGER_API_PATH}${type}/${endpoint}" > "${SWAGGER_SPEC_OUT_DIR}/${type}-${endpoint}.json"` exited with status 35.

[INFO] 		Stack Trace: 

[INFO] 		  1: /home/travis/gopath/src/github.com/openshift/origin/hack/update-generated-swagger-spec.sh:50: `curl -w "\n" "${SWAGGER_API_PATH}${type}/${endpoint}" > "${SWAGGER_SPEC_OUT_DIR}/${type}-${endpoint}.json"`

[INFO]   Exiting with code 35.

[INFO] Dumping etcd contents to /tmp/openshift/generate-swagger-spec//artifacts/etcd_dump.json

[INFO] Tearing down test

[INFO] Pruning etcd data directory...

/home/travis/gopath/src/github.com/openshift/origin/hack/util.sh: line 393: journalctl: command not found

[INFO] Cleanup complete

[FAIL] !!!!! Generate Failed !!!!
```

dcbw commented:
Travis failures seem to consistently be https://github.com/openshift/origin/issues/11517

smarterclayton commented:
@liggitt the TLS change broke travis curl

smarterclayton commented:
Actually...

@liggitt


On Mon, Oct 24, 2016 at 12:41 AM, Dan Williams <notifications@github.com>
wrote:

> travis flake:
>
> INFO] Updating /home/travis/gopath/src/github.com/openshift/origin/_output/verify-generated-swagger-spec/api/swagger-spec/oapi-v1.json from https://127.0.0.1:38443/swaggerapi/oapi/v1...
>
>   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
>
>                                  Dload  Upload   Total   Spent    Left  Speed
>
>   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0curl: (35) error:1407742E:SSL routines:SSL23_GET_SERVER_HELLO:tlsv1 alert protocol version
>
> [ERROR] PID 23055: /home/travis/gopath/src/github.com/openshift/origin/hack/update-generated-swagger-spec.sh:50: `curl -w "\n" "${SWAGGER_API_PATH}${type}/${endpoint}" > "${SWAGGER_SPEC_OUT_DIR}/${type}-${endpoint}.json"` exited with status 35.
>
> [INFO]      Stack Trace:
>
> [INFO]        1: /home/travis/gopath/src/github.com/openshift/origin/hack/update-generated-swagger-spec.sh:50: `curl -w "\n" "${SWAGGER_API_PATH}${type}/${endpoint}" > "${SWAGGER_SPEC_OUT_DIR}/${type}-${endpoint}.json"`
>
> [INFO]   Exiting with code 35.
>
> [INFO] Dumping etcd contents to /tmp/openshift/generate-swagger-spec//artifacts/etcd_dump.json
>
> [INFO] Tearing down test
>
> [INFO] Pruning etcd data directory...
>
> /home/travis/gopath/src/github.com/openshift/origin/hack/util.sh: line 393: journalctl: command not found
>
> [INFO] Cleanup complete
>
> [FAIL] !!!!! Generate Failed !!!!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/11082#issuecomment-255646700>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_pzYN-z7Z4r37O7DRDNaSMiq99nVxks5q3DbfgaJpZM4KFZTf>
> .
>

smarterclayton commented:
Good old curl

On Mon, Oct 24, 2016 at 1:10 AM, Clayton Coleman <ccoleman@redhat.com>
wrote:

> Actually...
>
> @liggitt
>
>
> On Mon, Oct 24, 2016 at 12:41 AM, Dan Williams <notifications@github.com>
> wrote:
>
>> travis flake:
>>
>> INFO] Updating /home/travis/gopath/src/github.com/openshift/origin/_output/verify-generated-swagger-spec/api/swagger-spec/oapi-v1.json from https://127.0.0.1:38443/swaggerapi/oapi/v1...
>>
>>   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
>>
>>                                  Dload  Upload   Total   Spent    Left  Speed
>>
>>   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0curl: (35) error:1407742E:SSL routines:SSL23_GET_SERVER_HELLO:tlsv1 alert protocol version
>>
>> [ERROR] PID 23055: /home/travis/gopath/src/github.com/openshift/origin/hack/update-generated-swagger-spec.sh:50: `curl -w "\n" "${SWAGGER_API_PATH}${type}/${endpoint}" > "${SWAGGER_SPEC_OUT_DIR}/${type}-${endpoint}.json"` exited with status 35.
>>
>> [INFO]      Stack Trace:
>>
>> [INFO]        1: /home/travis/gopath/src/github.com/openshift/origin/hack/update-generated-swagger-spec.sh:50: `curl -w "\n" "${SWAGGER_API_PATH}${type}/${endpoint}" > "${SWAGGER_SPEC_OUT_DIR}/${type}-${endpoint}.json"`
>>
>> [INFO]   Exiting with code 35.
>>
>> [INFO] Dumping etcd contents to /tmp/openshift/generate-swagger-spec//artifacts/etcd_dump.json
>>
>> [INFO] Tearing down test
>>
>> [INFO] Pruning etcd data directory...
>>
>> /home/travis/gopath/src/github.com/openshift/origin/hack/util.sh: line 393: journalctl: command not found
>>
>> [INFO] Cleanup complete
>>
>> [FAIL] !!!!! Generate Failed !!!!
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/openshift/origin/pull/11082#issuecomment-255646700>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/ABG_pzYN-z7Z4r37O7DRDNaSMiq99nVxks5q3DbfgaJpZM4KFZTf>
>> .
>>
>
>

akostadinov commented on pkg/sdn/plugin/node_iptables.go:
```diff
@@ -91,7 +91,7 @@ func (n *NodeIPTables) syncIPTableRules() error {
 func (n *NodeIPTables) getStaticNodeIPTablesRules() []FirewallRule {
 	return []FirewallRule{
-		{"nat", "POSTROUTING", []string{"-s", n.clusterNetworkCIDR, "!", "-d", n.clusterNetworkCIDR, "-j", "MASQUERADE"}},
+		{"nat", "POSTROUTING", []string{"-s", n.clusterNetworkCIDR, "-j", "MASQUERADE"}},
```
why did we change this?

danwinship replied:
Before, when we used docker for IPAM, docker's own iptables rules ended up applying to us, because we were configuring docker to use the same IP range as OpenShift. Since we're no longer reconfiguring docker's networking, we no longer get those rules, so we need to change our rules a bit to compensate.

In particular, we need a rule that says that if iptables sees traffic going from one pod to another pod, then it needs to masquerade it. That's because the only time that pod-to-pod traffic would end up being seen by iptables is if the traffic was originally pod-IP-to-service-IP, and then iptables DNATed the service IP to a pod IP. In that case we need to masquerade, because if we don't, then the destination pod would try to send its response directly back to the source pod, which would happen on the OVS bridge without iptables ever seeing it, and so the DNATing wouldn't get reversed, and so the source pod wouldn't recognize the packet. By masquerading it to the node's IP, we ensure that the service's response gets bounced out of OVS back to where iptables can see it again so it can un-DNAT it before passing it back to the source pod.

Since we already had a rule saying that pod-to-external traffic needs to be masqueraded, and we also need a rule saying that pod-to-pod traffic (that gets seen by iptables) needs to be masqueraded, we can just squash those rules together to get what you see here: any traffic from a pod that ends up being seen by iptables needs to be masqueraded.

akostadinov replied:
Understood, thank you for thorough and clear explanation.
